<h1 align=center style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
    بسکتبالیست‌ها
</font>
</h1>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
        در این سوال قصد داریم تا داده‌های مربوط به عملکرد بازیکنان لیگ بسکتبال NBA را تحلیل کنیم.
        پس کافی است تا کار‌هایی که از شما خواسته شده را مرحله به مرحله انجام دهید.
    </font>
</p>

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        مجموعه‌داده
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
        مجموعه‌داده‌ای که در احتیار شما قرار گرفته است شامل اطلاعات عملکرد بازیکنان NBA در فصل‌های ۲۰۰۰ الی ۲۰۰۹ است.
    این مجموعه شامل ۱۸۳۰ سطر و ۱۸ ستون می‌باشد که هر سطر آن شامل اطلاعات یک بازیکن در یک فصل است. توضیحات مربوط به ستون‌ها نیز در جدول زیر آمده است.
    </font>
</p>

<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|ستون|توضیحات|
|:------:|:---:|
|<code>PlAYER</code>|نام بازیکن|
|<code>TEAM</code>|مخفف اسم تیم بازیکن در آن فصل|
|<code>YEAR</code>|سال (فصل مسابقات)|
|<code>GP</code>|تعداد بازی‌های انجام شده|
|<code>MIN</code>|میانگین دقایق بازی کرده|
|<code>PTS</code>|میانگین امتیاز|
|<code>FGM</code>|میانگین تعداد پرتاب‌های ۲ امتیازی موفق|
|<code>FGA</code>|میانگین تعداد کل پرتاب‌های ۲ امتیازی|
|<code>3PM</code>|میانگین تعداد پرتاب‌های ۳ امتیازی موفق|
|<code>3PA</code>|میانگین تعداد کل پرتاب‌های ۳ امتیازی|
|<code>FTM</code>|میانگین تعداد پرتاب‌های آزاد (پنالتی) موفق|
|<code>FTA</code>|میانگین تعداد کل پرتاب‌های آزاد (پنالتی)|
|<code>OREB</code>|میانگین تعداد ریباند‌ها در حمله|
|<code>DREB</code>|میانگین تعداد ریباند‌ها در دفاع|
|<code>AST</code>|میانگین تعداد پاس‌های منجر به امتیاز|
|<code>STL</code>|میانگین تعداد دفعات توپ‌ربایی|
|<code>BLK</code>|میانگین تعداد بلاک‌ها|
|<code>TOV</code>|میانگین تعداد دفعات لو دادن توپ|
    
</font>
</div>
</center>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
   با استفاده از سلول زیر کتابخانه‌های مورد نیاز خود را <code>import</code> کنید و سپس مجموعه‌داده را که در فایلی به نام <code>NBA2000-2009.csv</code> قرار دارد، بخوانید. 
</font>
</p>

In [37]:
import pandas as pd
import math

df = pd.read_csv('../Data/NBA2000-2009.csv')
df.head()

PLAYER TEAM  YEAR  GP   MIN   PTS   FGM   FGA  3PM  3PA  FTM  \
0     Allen Iverson  PHI  2000  71  42.0  31.1  10.7  25.5  1.4  4.3  8.2   
1  Jerry Stackhouse  DET  2000  80  40.2  29.8   9.7  24.1  2.1  5.9  8.3   
2  Shaquille O'Neal  LAL  2000  74  39.5  28.7  11.0  19.2  0.0  0.0  6.7   
3       Kobe Bryant  LAL  2000  68  40.9  28.5  10.3  22.2  0.9  2.9  7.0   
4      Vince Carter  TOR  2000  75  39.7  27.6  10.2  22.1  2.2  5.3  5.1   

    FTA  OREB  DREB  AST  STL  BLK  TOV  
0  10.1   0.7   3.1  4.6  2.5  0.3  3.3  
1  10.1   1.2   2.7  5.1  1.2  0.7  4.1  
2  13.1   3.9   8.8  3.7  0.6  2.8  2.9  
3   8.2   1.5   4.3  5.0  1.7  0.6  3.2  
4   6.7   2.3   3.2  3.9  1.5  1.1  2.2

In [281]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1830 entries, 0 to 1829
Data columns (total 18 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   PLAYER  1830 non-null   object 
 1   TEAM    1830 non-null   object 
 2   YEAR    1830 non-null   int64  
 3   GP      1830 non-null   int64  
 4   MIN     1830 non-null   float64
 5   PTS     1830 non-null   float64
 6   FGM     1830 non-null   float64
 7   FGA     1830 non-null   float64
 8   3PM     1830 non-null   float64
 9   3PA     1830 non-null   float64
 10  FTM     1830 non-null   float64
 11  FTA     1830 non-null   float64
 12  OREB    1830 non-null   float64
 13  DREB    1830 non-null   float64
 14  AST     1830 non-null   float64
 15  STL     1830 non-null   float64
 16  BLK     1830 non-null   float64
 17  TOV     1830 non-null   float64
dtypes: float64(14), int64(2), object(2)
memory usage: 257.5+ KB


<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت اول
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         لیستی از بازیکن‌ها و فصل‌ها را بدست آورید که در آن فصل، آن بازیکن در حداقل ۲ تا از ۵ ملاک اصلی آمار دو رقمی (بزرگتر مساوی ۱۰) ثبت کرده باشد. 
        <br>
        منظور از ۵ ملاک اصلی میانگین امتیاز، میانگین تعداد ریباند (مجموع ریباند در دفاع و حمله)، میانگین تعداد پاس‌های منجر به گل، میانگین تعداد بلاک‌ها و میانگین تعداد توپ‌ربایی‌ها است.
    </font>
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>double_double</code> قرار دهید.
    این دیتافریم باید شامل ۷ ستون باشد که ستون‌ها از چپ به راست به ترتیب نام بازیکن، سال، میانگین امتیاز، میانگین تعداد پاس‌های منجر به گل، میانگین تعداد ریباند‌ها، میانگین تعداد بلاک‌ها و میانگین تعداد توپ‌ربایی‌ها باشند. همچنین کل دیتافریم باید به ترتیب برحسب سال و نام بازیکن به صورت صعودی مرتب باشد. دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|PLAYER|YEAR|PTS|AST|REB|STL|BLK|
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|Antonio Davis|2000|13.7|1.4|10.1|0.3|1.9|
|Antonio McDyess|2000|20.8|2.1|12.0|0.6|1.5|
    
</font>
</div>
</center>

In [48]:
df['REB'] = df['OREB'] + df['DREB']


a = df.loc[:,['PLAYER','YEAR','PTS','AST','REB','STL','BLK']]

col = ['PTS','AST','REB','STL','BLK']
condition = df[col] >= 10


greater_than_ten = a[condition.sum(axis=1) >= 2]

double_double = greater_than_ten.sort_values(by=['YEAR', 'PLAYER'], ascending=True)
# double_double[col]






<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت دوم
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         باارزش‌ترین بازیکنان این ۱۰ فصل را مشخص کنید.
        ارزش یک بازیکن در یک فصل با استفاده از فرمول زیر بدست می‌آید:
    </font>
    $$ (PTS + OREB + DREB + AST + STL + BLK) - (TOV + Missed FG + Missed FT) $$
    منظور از <code>Missed FG</code> و <code>Missed FT</code> به ترتیب میانگین تعداد پرتاب‌های ۲ امتیازی و آزاد از دست رفته است.  
    <br>
    ارزش یک بازیکن در کل ۱۰ فصل برابر است با میانگین ارزش او در هر فصل.
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>best_player</code> قرار دهید.
    این دیتافریم شامل دو ستون است که یکی نام بازیکن و دومی ارزش آن بازیکن است که تا دو رقم اعشار باید گرد شود. همچنین این دیتافریم باید بر اساس ارزش به صورت نزولی و سپس بر اساس نام بازیکن به‌صورت صعودی مرتب باشد.
    دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|PLAYER|VALUE|
|:------:|:---:|
|Kevin Garnett|29.74|
|LeBron James|27.9|
    
</font>
</div>
</center>

In [102]:
double_double.reset_index(drop=True, inplace=True)

double_double['sum_score'] = double_double[col].sum(axis = 1)
double_double['final_score'] = (double_double.sum_score) - (df['TOV']+df['FGA']+df['FTA'])


# grouped_score_in_year = double_double.groupby(['PLAYER', 'YEAR'])['final_score'].mean()
grouped_score = double_double.groupby('PLAYER')['final_score'].mean().reset_index()
grouped_score = grouped_score.rename(columns = {'final_score':'VALUE'})
grouped_score
best_player = pd.DataFrame(grouped_score)


best_player['VALUE'] = best_player['VALUE'].apply(lambda x: math.ceil(x * 100) / 100)






# best_player['PLAYER'] = double_double['PLAYER']
# best_player['VALUE'] = double_double['VALUE']

# best_player = grouped_score_in_year.reset_index()


# best_player = best_player.drop(columns=['YEAR'])

# best_player.columns = ['PLAYER', 'VALUE']










In [104]:
best_player

PLAYER  VALUE
0       Al Jefferson  18.30
1       Andre Miller   5.20
2     Antawn Jamison  18.60
3      Antonio Davis -11.50
4    Antonio McDyess  -1.29
5        Brad Miller   6.90
6        Brian Grant  -1.99
7      Carlos Boozer  16.33
8         Chris Bosh  23.04
9         Chris Paul  26.25
10      Chris Webber  13.45
11     Danny Fortson  -3.89
12         David Lee  18.75
13    Deron Williams  19.81
14   Dikembe Mutombo  -3.90
15     Dirk Nowitzki  10.70
16     Dwight Howard  16.75
17       Elton Brand  11.14
18      Emeka Okafor  11.14
19     Erick Dampier   3.80
20    Gerald Wallace  19.70
21   Jamaal Magloire   5.10
22        Jason Kidd  15.20
23   Jermaine O'Neal  11.37
24      Kenny Thomas   5.20
25     Kevin Garnett  20.05
26       Kurt Thomas   3.40
27        Lamar Odom  15.50
28      Marcus Camby  11.90
29  Samuel Dalembert   7.90
30  Shaquille O'Neal  18.14
31      Shawn Marion  11.87
32        Steve Nash  15.69
33        Tim Duncan  18.89
34       Troy Murphy   8.88
35    Tyson Chandler   9.21
36     Zach Randolph  18.34

In [96]:
# grouped_score_in_year


In [110]:
best_player = best_player.sort_values(by=['VALUE', 'PLAYER'], ascending=[False, True])

best_player



PLAYER  VALUE
9         Chris Paul  26.25
8         Chris Bosh  23.04
25     Kevin Garnett  20.05
13    Deron Williams  19.81
20    Gerald Wallace  19.70
33        Tim Duncan  18.89
12         David Lee  18.75
2     Antawn Jamison  18.60
36     Zach Randolph  18.34
0       Al Jefferson  18.30
30  Shaquille O'Neal  18.14
16     Dwight Howard  16.75
7      Carlos Boozer  16.33
32        Steve Nash  15.69
27        Lamar Odom  15.50
22        Jason Kidd  15.20
10      Chris Webber  13.45
28      Marcus Camby  11.90
31      Shawn Marion  11.87
23   Jermaine O'Neal  11.37
17       Elton Brand  11.14
18      Emeka Okafor  11.14
15     Dirk Nowitzki  10.70
35    Tyson Chandler   9.21
34       Troy Murphy   8.88
29  Samuel Dalembert   7.90
5        Brad Miller   6.90
1       Andre Miller   5.20
24      Kenny Thomas   5.20
21   Jamaal Magloire   5.10
19     Erick Dampier   3.80
26       Kurt Thomas   3.40
4    Antonio McDyess  -1.29
6        Brian Grant  -1.99
11     Danny Fortson  -3.89
14   Dikembe Mutombo  -3.90
3      Antonio Davis -11.50

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت سوم
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         در هر سال چه تیمی بیشترین میانگین امتیاز را ثبت کرده است؟
        <br>
        توجه داشته باشید که میانگین امتیاز یک تیم برابر است با مجموع میانگین امتیاز بازیکن‌هایش.
    </font>
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>max_PTS_of_year</code> قرار دهید.
    این دیتافریم باید شامل سه ستون باشد که ستون‌ها از چپ به راست به ترتیب سال، نام تیم و میانگین امتیاز است که امتیاز‌ها باید تا دو رقم اعشار گرد شوند. همچنین کل دیتافریم باید برحسب سال به صورت صعودی مرتب باشد. دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|YEAR|TEAM|PTS|
|:------:|:---:|:---:|
|2000|SAC|100.2|
|2001|LAL|95.9|
    
</font>
</div>
</center>

In [ ]:
max_PTS_of_year = #TODO


<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
<b>سلول جواب‌ساز</b>
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
    برای ساخته‌شدن فایل <code>result.zip</code> سلول زیر را اجرا کنید. توجه داشته باشید که پیش از اجرای سلول زیر تغییرات اعمال شده در نت‌بوک را ذخیره کرده باشید (<code>ctrl+s</code>) در غیر این صورت، در پایان مسابقه نمره شما به صفر تغییر خواهد کرد.
    <br>
    همچنین اگر از کولب برای اجرای این فایل نوت‌بوک استفاده می‌کنید، قبل از ارسال فایل <code>result.zip</code>، آخرین نسخه‌ی نوت‌بوک خود را دانلود کرده و داخل فایل ارسالی قرار دهید.
</font>

In [ ]:
import zipfile

if not os.path.exists(os.path.join(os.getcwd(), 'NBA.ipynb')):
    %notebook -e NBA.ipynb

double_double.to_csv('double_double.csv', index=False)
best_player.to_csv('best_player.csv', index=False)
max_PTS_of_year.to_csv('max_PTS_of_year.csv', index=False)

def compress(file_names):
    print("File Paths:")
    print(file_names)
    compression = zipfile.ZIP_DEFLATED
    with zipfile.ZipFile("result.zip", mode="w") as zf:
        for file_name in file_names:
            zf.write('./' + file_name, file_name, compress_type=compression)

file_names = ['double_double.csv', 'best_player.csv', 'max_PTS_of_year.csv', 'NBA.ipynb']
compress(file_names)